src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -345,6 +345,12 @@ private bool TryClose()
 
         public async Task ProcessRequestAsync<TContext>(IHttpApplication<TContext> application) where TContext : notnull
         {
+            // Start request header timeout. Reset when headers are received by a request stream.
+            if (TimeoutControl.TimerReason == TimeoutReason.None)
```

Comment:
Note that this is setting a shared connection timeout control on a per-stream basis. This doesn't seem very threadsafe. I wanted to get thoughts on the best way to do this before improve it.

Should it be used in a lock, or should there be a timeout control per stream?

Reply:
Probably a timeout control per stream. Timeouts in HTTP/3 now have IO on each stream, and the RequestHeadersTimeout should be specific to a stream.